File: scripts/analyze.py
#!/usr/bin/env python3
"""
code-map Phase 1: mechanical structural extraction.

Walks the project, dispatches each source file to the appropriate
tree-sitter extractor, builds the dependency graph, assigns layers,
and writes raw_structure.json.

Phase 2 (AI refinement in the slash command) consumes this and produces
the human-friendly code-map.json with descriptions and layer overrides.
"""
from __future__ import annotations
import argparse
import datetime as dt
import json
import os
import sys
from collections import Counter, defaultdict
from pathlib import Path


# Make the `scripts.lib.*` package import work regardless of cwd.
HERE = Path(__file__).resolve().parent
ROOT_OF_PACKAGE = HERE.parent
if str(ROOT_OF_PACKAGE) not in sys.path:
    sys.path.insert(0, str(ROOT_OF_PACKAGE))

# Also expose the bootstrap-installed grammars (if any) on sys.path
_DATA = os.environ.get("CLAUDE_PLUGIN_DATA") or str(Path.home() / ".cache" / "code-map")
_WHEELS = Path(_DATA) / "wheels"
if _WHEELS.exists() and str(_WHEELS) not in sys.path:
    sys.path.insert(0, str(_WHEELS))


def walk_project(root: Path, skip_dirs: set[str]):
    """Yield Path objects for all source files we might extract.

    Uses os.walk with in-place dir pruning so we never descend into skipped
    trees (node_modules, build, testsuites, …). rglob('*') + a post-filter
    would still enumerate and stat() everything inside those dirs.
    """
    from scripts.lib.extractors import all_extensions
    from scripts.lib.skipdirs import prune_dirnames
    exts = all_extensions()
    for dirpath, dirnames, filenames in os.walk(root):
        dirnames[:] = prune_dirnames(dirnames, skip_dirs, filenames)
        for fn in filenames:
            if Path(fn).suffix in exts:
                yield Path(dirpath) / fn


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--root", default=".", help="Project root")
    ap.add_argument("--out", default=".code-map/raw_structure.json",
                    help="Output JSON path")
    ap.add_argument("--core-percentile", type=float, default=0.25,
                    help="Top fraction per layer to mark as core (default 0.25)")
    ap.add_argument("--core-max-per-layer", type=int, default=40,
                    help="Absolute cap on core declarations per layer (0 = no cap, default 40)")
    ap.add_argument("--flow-hub-percentile", type=float, default=0.05,
                    help="Top fraction by global in-degree to treat as flow hubs "
                         "(non-expandable leaves; default 0.05)")
    ap.add_argument("--flow-max-depth", type=int, default=6,
                    help="Max hops a flow expands from its entry point (default 6)")
    ap.add_argument("--skip", action="append", default=None, metavar="DIR",
                    help="Extra directory name to skip (repeatable); merges with the defaults "
                         "and .code-map/skip-dirs.txt")
    ap.add_argument("--name", default=None, help="Project display name")
    ap.add_argument("--detect-only", action="store_true",
                    help="Run only template detection; write detection.json next "
                         "to --out and skip extraction (used by Phase 0)")
    args = ap.parse_args()

    root = Path(args.root).resolve()
    out_path = Path(args.out)
    if not out_path.is_absolute():
        out_path = root / out_path
    out_path.parent.mkdir(parents=True, exist_ok=True)

    # Imports must come after the sys.path setup above.
    from scripts.lib.extractors import extractor_for
    from scripts.lib import core, layers, flows as flowmod, gitmeta, vendoring
    from scripts.lib.skipdirs import load_skip_dirs

    # plugin_root resolves to this script's grandparent (the code-map repo).
    # When the plugin is installed normally, $CLAUDE_PLUGIN_ROOT also points here.
    plugin_root = Path(os.environ.get("CLAUDE_PLUGIN_ROOT") or HERE.parent)
    layer_config, detection = layers.load_config(root, plugin_root)

    if args.detect_only:
        detect_path = out_path.parent / "detection.json"
        with open(detect_path, "w") as f:
            json.dump(detection, f, indent=2)
        if detection.get("reason"):
            print(f"[analyze] template: {detection['chosen']} (fallback: {detection['reason']})")
        else:
            ranked = sorted(detection["scores"].items(), key=lambda kv: kv[1], reverse=True)[:3]
            ranked_str = ", ".join(f"{tid}={sc}" for tid, sc in ranked)
            print(f"[analyze] template: {detection['chosen']} (top: {ranked_str})")
        print(f"[analyze] wrote {detect_path} (detect-only; skipped extraction)")
        return

    skip_dirs = load_skip_dirs(root, args.skip)
    files = list(walk_project(root, skip_dirs))
    all_decls = []
    all_skipped = []
    parse_failures = 0
    lang_counts = Counter()
    files_by_lang = defaultdict(int)

    for path in files:
        ext = path.suffix
        extractor = extractor_for(ext)
        if extractor is None:
            continue
        try:
            src = path.read_bytes()
            result = extractor.parse(path, src, root)
            for d in result.declarations:
                d.language = extractor.name
                all_decls.append(d)
            all_skipped.extend(result.skipped)
            lang_counts[extractor.name] += len(result.declarations)
            files_by_lang[extractor.name] += 1
        except Exception as e:
            parse_failures += 1
            all_skipped.append({"path": str(path.relative_to(root)),
                                "reason": f"exception: {type(e).__name__}: {e}"})

    # Build graph
    decls, edges = core.build_graph(all_decls)

    # Assign layers and core flag
    layers.apply_to(decls, layer_config)
    core.mark_core(decls, percentile=args.core_percentile,
                   max_per_layer=args.core_max_per_layer)

    # Flows: mark high-in-degree hubs, then build one candidate flow per entry point.
    hub_ids = flowmod.mark_hubs(decls, percentile=args.flow_hub_percentile)
    entry_seeds = [d.qualified_name for d in decls if core.is_entry_point(d)]
    flow_list = flowmod.build_flows(entry_seeds, decls, edges, hub_ids,
                                    max_depth=args.flow_max_depth)

    # Vendored-flooding advisory: flag top-level dirs dominated by packages
    # outside the project's own roots (likely vendored toolchains) so Phase 2 /
    # the user can skip them. Advisory only — never changes extraction. Own roots
    # come from build manifests (what the project declares as its own); entry
    # points are an unreliable fallback here because vendored toolchains are full
    # of Main/*Application classes, so use them only if no manifest declared one.
    manifest_names = ("build.gradle", "build.gradle.kts", "pom.xml")
    manifest_dirs = [root] + [p for p in sorted(root.iterdir())
                              if p.is_dir() and p.name not in skip_dirs]
    manifest_texts = []
    for base in manifest_dirs:
        for mn in manifest_names:
            mf = base / mn
            if mf.is_file():
                try:
                    manifest_texts.append(mf.read_text(errors="ignore"))
                except OSError:
                    pass
    own_roots = vendoring.extract_own_roots(manifest_texts)
    if not own_roots:
        own_roots = {vendoring.package_root(d.namespace)
                     for d in decls if core.is_entry_point(d)}
        own_roots.discard("")
    advisories = vendoring.detect_vendored_dirs(
        [(d.path.split("/", 1)[0], vendoring.package_root(d.namespace))
         for d in decls],
        own_roots)

    project_meta = {
        "name": args.name or root.name,
        "root": str(root),
        "languages": sorted(lang_counts.keys()),
        "files_scanned": len(files),
        "files_by_language": dict(files_by_lang),
        "declarations_by_language": dict(lang_counts),
        "parse_failures": parse_failures,
        "generated_at": dt.datetime.now().isoformat(timespec="seconds"),
    }
    # Always present so Phase 2 (build.md step 0) can rely on it. On the
    # fallback paths it carries a `reason` and empty scores/evidence instead
    # of detection signals — see layers.load_config.
    project_meta["template_detection"] = detection
    if advisories:
        project_meta["advisories"] = advisories  # present only when something flagged
    git = gitmeta.git_info(root)
    if git:
        project_meta["git"] = git          # omitted entirely when not a git repo

    data = core.to_json_shape(
        declarations=decls,
        edges=edges,
        layers=[{k: l[k] for k in ("id", "name", "order", "summary")} for l in layer_config],
        project_meta=project_meta,
        flows=flow_list,
    )

    with open(out_path, "w") as f:
        json.dump(data, f, indent=2, default=str)

    # Write unresolved.json so Phase 2 AI step has a triage list
    unresolved_path = out_path.parent / "unresolved.json"
    with open(unresolved_path, "w") as f:
        json.dump({
            "skipped": all_skipped,
            "low_confidence": [
                {"id": d.qualified_name, "path": d.path, "reason": "low_confidence"}
                for d in decls if d.confidence != "high"
            ]
        }, f, indent=2)

    print(f"[analyze] root: {root}")
    print(f"[analyze] languages: {', '.join(project_meta['languages']) or '(none)'}")
    if detection.get("reason"):
        print(f"[analyze] template: {detection['chosen']} (fallback: {detection['reason']} "
              f"— no signal-based detection; Phase 2 should verify the architecture)")
    else:
        ranked = sorted(detection["scores"].items(), key=lambda kv: kv[1], reverse=True)[:3]
        ranked_str = ", ".join(f"{tid}={sc}" for tid, sc in ranked)
        print(f"[analyze] template: {detection['chosen']} (top: {ranked_str})")
    print(f"[analyze] files scanned: {len(files)}  declarations: {len(decls)}  edges: {len(edges)}")
    print(f"[analyze] flows: {len(flow_list)} (entry-point seeds)")
    print(f"[analyze] skipped/low-confidence: {len(all_skipped)} entries")
    for a in advisories:
        print(f"[analyze] advisory: '{a['dir']}/' has {a['declarations']} decls, "
              f"{a['foreign_pct']}% outside your code ({a['dominant_foreign']}.*) — "
              f"likely vendored; add '{a['dir']}' to .code-map/skip-dirs.txt to focus the map")
    print(f"[analyze] wrote {out_path}")
    print(f"[analyze] wrote {unresolved_path}")


if __name__ == "__main__":
    main()

File: scripts/analyze.mjs
// scripts/analyze.mjs — code-map Phase 1: mechanical structural extraction.
// Port of analyze.py. Walks the project, dispatches each source file to a
// web-tree-sitter extractor, builds the graph, assigns layers, writes
// raw_structure.json (+ unresolved.json). Deterministic — never guesses.
import { readFileSync, readdirSync, writeFileSync, mkdirSync, statSync, realpathSync } from 'node:fs';
import { resolve as resolvePath, relative, join, sep, basename, dirname, isAbsolute } from 'node:path';
import { fileURLToPath } from 'node:url';

import { allExtensions, extractorFor } from './lib/extractors/index.mjs';
import { qualifiedName } from './lib/extractors/base.mjs';
import { assignDisplayNames } from './lib/labels.mjs';
import * as core from './lib/core.mjs';
import * as layers from './lib/layers.mjs';
import * as flowmod from './lib/flows.mjs';
import * as gitmeta from './lib/gitmeta.mjs';
import * as vendoring from './lib/vendoring.mjs';
import { loadSkipDirs, pruneDirnames } from './lib/skipdirs.mjs';
import { GrammarUnavailable } from './lib/grammars.mjs';
import { pluginVersion, codeMapFingerprints } from './lib/version.mjs';
import { resolveProject } from './lib/resolve/index.mjs';

const HERE = dirname(fileURLToPath(import.meta.url));

export function parseArgs(argv) {
  const a = {
    root: '.', out: '.code-map/raw_structure.json',
    core_percentile: 0.3, core_max_per_layer: 40, core_min_per_layer: 4,
    flow_hub_percentile: 0.05, flow_max_depth: 8, flow_seed_max: 12, flow_max_nodes: 60,
    skip: [], name: null,
    extract_only: false, layer_only: false, extract: '.code-map/extract.json',
  };
  for (let i = 0; i < argv.length; i++) {
    const k = argv[i];
    const next = () => argv[++i];
    if (k === '--root') a.root = next();
    else if (k === '--out') a.out = next();
    else if (k === '--core-percentile') a.core_percentile = parseFloat(next());
    else if (k === '--core-max-per-layer') a.core_max_per_layer = parseInt(next(), 10);
    else if (k === '--core-min-per-layer') a.core_min_per_layer = parseInt(next(), 10);
    else if (k === '--flow-hub-percentile') a.flow_hub_percentile = parseFloat(next());
    else if (k === '--flow-max-depth') a.flow_max_depth = parseInt(next(), 10);
    else if (k === '--flow-seed-max') a.flow_seed_max = parseInt(next(), 10);
    else if (k === '--flow-max-nodes') a.flow_max_nodes = parseInt(next(), 10);
    else if (k === '--skip') a.skip.push(next());
    else if (k === '--name') a.name = next();
    else if (k === '--extract-only') a.extract_only = true;
    else if (k === '--layer-only') a.layer_only = true;
    else if (k === '--extract') a.extract = next();
  }
  return a;
}

function isDir(p) { try { return statSync(p).isDirectory(); } catch { return false; } }
function isFile(p) { try { return statSync(p).isFile(); } catch { return false; } }

// Yield source-file paths under root, os.walk-style top-down with in-place prune.
// Entries are visited in sorted order (deterministic walk) and every yielded file
// is deduped by realpath, so a physical file reachable through more than one path
// (SwiftPM repos symlink a shared `Platform/` module into each target dir) is
// parsed exactly once — the first, lexicographically-earliest path wins.
export function* walkProject(root, skipDirs) {
  const exts = allExtensions();
  const seenReal = new Set();
  const walk = function* (dir) {
    let entries;
    try { entries = readdirSync(dir, { withFileTypes: true }); } catch { return; }
    const dirnames = entries.filter((e) => e.isDirectory()).map((e) => e.name).sort();
    const filenames = entries.filter((e) => !e.isDirectory()).map((e) => e.name).sort();
    for (const fn of filenames) {
      const dot = fn.lastIndexOf('.');
      const ext = dot >= 0 ? fn.slice(dot) : '';
      if (!exts.has(ext)) continue;
      const full = join(dir, fn);
      let real;
      try { real = realpathSync(full); } catch { continue; } // broken symlink → unreadable
      if (seenReal.has(real)) continue; // same physical file via another path
      seenReal.add(real);
      yield full;
    }
    for (const d of pruneDirnames(dirnames, skipDirs, filenames)) yield* walk(join(dir, d));
  };
  yield* walk(root);
}

// 全树扩展名普查 —— 与 walkProject 同样的遍历/剪枝,但统计所有文件扩展名
// (含无 extractor 的),供 dominantUnsupportedLanguage 护栏使用。
function surveyExtensions(root, skipDirs) {
  const counts = new Map();
  const walk = (dir) => {
    let entries;
    try { entries = readdirSync(dir, { withFileTypes: true }); } catch { return; }
    const dirnames = entries.filter((e) => e.isDirectory()).map((e) => e.name).sort();
    const filenames = entries.filter((e) => !e.isDirectory()).map((e) => e.name).sort();
    for (const fn of filenames) {
      const dot = fn.lastIndexOf('.');
      const ext = dot >= 0 ? fn.slice(dot) : '';
      if (ext) counts.set(ext, (counts.get(ext) || 0) + 1);
    }
    for (const d of pruneDirnames(dirnames, skipDirs, filenames)) walk(join(dir, d));
  };
  walk(root);
  return counts;
}

function extOf(p) { const b = basename(p); const i = b.lastIndexOf('.'); return i >= 0 ? b.slice(i) : ''; }
function rel(root, p) { return relative(root, p).split(sep).join('/'); }

// 常见「代码」扩展 → 语言名,仅收录**没有 extractor** 的语言(用于无支持语言护栏)。
// 受支持语言的扩展(.py/.go/.ts/.c/... )刻意不在表内。
const KNOWN_CODE_EXT = {
  '.sh': 'shell', '.bash': 'shell', '.zsh': 'shell', '.ksh': 'shell',
  '.rb': 'ruby', '.rake': 'ruby', '.gemspec': 'ruby',
  '.php': 'php',
  '.pl': 'perl', '.pm': 'perl',
  '.scala': 'scala', '.sc': 'scala',
  '.ex': 'elixir', '.exs': 'elixir',
  '.hs': 'haskell',
  '.jl': 'julia',
};

// 纯判定:给定全树扩展名计数与受支持源文件总数,返回主导的「无支持语言」
// (其文件数严格大于受支持文件总数)或 null。确定性:按语言名升序遍历,取最大文件数者。
export function dominantUnsupportedLanguage(extCounts, supportedFileCount) {
  const byLang = new Map();
  for (const [ext, n] of extCounts) {
    const lang = KNOWN_CODE_EXT[ext];
    if (!lang) continue;
    byLang.set(lang, (byLang.get(lang) || 0) + n);
  }
  let top = null;
  for (const [language, files] of [...byLang].sort((a, b) => (a[0] < b[0] ? -1 : 1))) {
    if (!top || files > top.files) top = { language, files };
  }
  if (top && top.files > supportedFileCount) return top;
  return null;
}

export function buildContext(args) {
  const root = resolvePath(args.root);
  const pluginRoot = process.env.CLAUDE_PLUGIN_ROOT || join(HERE, '..');
  return { root, pluginRoot };
}

function buildNamespaceHistogram(decls) {
  const h = new Map();
  for (const d of decls) { const ns = d.namespace || '(none)'; h.set(ns, (h.get(ns) || 0) + 1); }
  return Object.fromEntries([...h.entries()].sort((a, b) => (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0)));
}

export function serializeExtract(model) {
  return {
    declarations: model.declarations,
    edges: model.edges,
    ingredients: model.ingredients,
    template_detection: model.detectionScores,
    namespace_histogram: model.namespace_histogram,
  };
}

export function deserializeExtract(obj) {
  return {
    declarations: obj.declarations || [],
    edges: obj.edges || [],
    ingredients: obj.ingredients || {},
    detectionScores: obj.template_detection || null,
    namespace_histogram: obj.namespace_histogram || {},
  };
}

function writeUnresolved(model, outDir) {
  writeFileSync(join(outDir, 'unresolved.json'), JSON.stringify({
    skipped: model.skipped,
    low_confidence: model.declarations.filter((d) => d.confidence !== 'high')
      .map((d) => ({ id: qualifiedName(d), path: d.path, reason: 'low_confidence' })),
    resolution: model.resolution.unresolved,
  }, null, 2));
}

function writeResolution(model, outDir) {
  writeFileSync(join(outDir, 'resolution.json'), JSON.stringify(model.resolution, null, 2));
}

// STAGE 1 — architecture-INDEPENDENT extraction. Walks, parses, builds the dep
// graph, scores importance, names display labels, computes advisories. Assigns
// NO layers / core / flows — those need the architecture (chosen in Phase 2).
export async function runExtract(args, ctx) {
  const { root, pluginRoot } = ctx;
  const skipDirs = loadSkipDirs(root, args.skip);
  const files = [...walkProject(root, skipDirs)];
  const allDecls = [];
  const allSkipped = [];
  const importsByFile = new Map();
  const mentionsByFile = new Map();
  const reexportsByFile = new Map();
  let parseFailures = 0;
  const langCounts = new Map();
  const filesByLang = new Map();

  for (const path of files) {
    const ext = extOf(path);
    const extractor = await extractorFor(ext);
    if (extractor == null) continue;
    const relPath = rel(root, path);
    try {
      const src = readFileSync(path, 'utf8');
      mentionsByFile.set(relPath, core.sourceTokens(src));
      const result = await extractor.parse(relPath, src, root);
      for (const d of result.declarations) { d.language = extractor.name; allDecls.push(d); }
      allSkipped.push(...result.skipped);
      importsByFile.set(relPath, result.imports);
      reexportsByFile.set(relPath, result.reexports || []);
      langCounts.set(extractor.name, (langCounts.get(extractor.name) || 0) + result.declarations.length);
      filesByLang.set(extractor.name, (filesByLang.get(extractor.name) || 0) + 1);
    } catch (e) {
      parseFailures++;
      const reason = e instanceof GrammarUnavailable
        ? `grammar_${extractor.name}_unavailable_offline`
        : `exception: ${e?.name || 'Error'}: ${e?.message || e}`;
      allSkipped.push({ path: relPath, reason });
    }
  }

  const resolution = resolveProject(allDecls, importsByFile, reexportsByFile, { root });
  const [decls, edges] = core.buildGraph(allDecls, { mentionsByFile });
  assignDisplayNames(decls); // R3: write _display_name on cross-module name collisions (layer-independent)

  // Vendored-flooding advisory (architecture-independent).
  const manifestNames = ['build.gradle', 'build.gradle.kts', 'pom.xml'];
  const childDirs = (() => {
    try { return readdirSync(root).sort().map((n) => join(root, n)).filter((p) => isDir(p) && !skipDirs.has(basename(p))); }
    catch { return []; }
  })();
  const manifestTexts = [];
  for (const base of [root, ...childDirs]) {
    for (const mn of manifestNames) {
      const mf = join(base, mn);
      if (isFile(mf)) { try { manifestTexts.push(readFileSync(mf, 'utf8')); } catch { /* ignore */ } }
    }
  }
  let ownRoots = vendoring.extractOwnRoots(manifestTexts);
  if (ownRoots.size === 0) {
    ownRoots = new Set(decls.filter((d) => core.isEntryPoint(d)).map((d) => vendoring.packageRoot(d.namespace)));
    ownRoots.delete('');
  }
  const advisories = vendoring.detectVendoredDirs(
    decls.map((d) => [d.path.split('/')[0], vendoring.packageRoot(d.namespace)]),
    ownRoots);
  const guardAdvisories = [];
  const dom = dominantUnsupportedLanguage(surveyExtensions(root, skipDirs), files.length);
  if (dom) {
    guardAdvisories.push({
      kind: 'unsupported-dominant-language',
      language: dom.language,
      files: dom.files,
      note: '主体语言无 extractor,地图仅含受支持语言,可能严重不完整',
    });
    console.log(`[analyze] advisory: dominant language '${dom.language}' (${dom.files} files) has no extractor — map may be misleading`);
  }
  const allAdvisories = [...advisories, ...guardAdvisories];

  const git = gitmeta.gitInfo(root);
  const cmVer = pluginVersion(pluginRoot);
  const fp = codeMapFingerprints(pluginRoot);
  const totalCand = resolution.stats.edges_resolved + resolution.stats.edges_unresolved;

  const ingredients = {
    name: args.name || basename(root),
    root: String(root),
    languages: [...langCounts.keys()].sort(),
    files_scanned: files.length,
    files_by_language: Object.fromEntries(filesByLang),
    declarations_by_language: Object.fromEntries(langCounts),
    parse_failures: parseFailures,
    resolution: { ...resolution.stats, coverage: totalCand ? core.round3(resolution.stats.edges_resolved / totalCand) : 1 },
    advisories: allAdvisories,
    git: git || null,
    code_map_version: cmVer || null,
    extract_version: fp.extract_version ?? null,
    refine_version: fp.refine_version ?? null,
  };

  return {
    declarations: decls,
    edges,
    ingredients,
    detectionScores: layers.detectOnly(root, pluginRoot),
    namespace_histogram: buildNamespaceHistogram(decls),
    skipped: allSkipped,
    resolution,
  };
}

// STAGE 2 — architecture-DEPENDENT layering. Consumes a runExtract model + the
// resolved architecture (architecture.yml via loadConfig), assigns layers, marks
// core (per layer), seeds/builds flows, and assembles the final raw_structure
// JSON. The projectMeta key order MUST match the legacy single-pass output.
export function runLayer(model, args, ctx) {
  const { root, pluginRoot } = ctx;
  const [layerConfig, detection] = layers.loadConfig(root, pluginRoot);
  const { leaves: layerLeaves, groups: layerGroups } = layers.expandGroups(layerConfig);
  const decls = model.declarations;
  const edges = model.edges;

  layers.applyTo(decls, layerLeaves);
  detection.fit = layers.templateFit(decls, layerLeaves);
  core.markCore(decls, args.core_percentile, args.core_max_per_layer, args.core_min_per_layer);

  const hubIds = flowmod.markHubs(decls, args.flow_hub_percentile);
  const dispatchIndex = flowmod.buildDispatchIndex(decls);
  const { seeds, seedKind } = flowmod.selectFlowSeeds(decls, { maxSeeds: args.flow_seed_max });
  let bfsFlows = flowmod.buildFlows(seeds, decls, edges, hubIds, args.flow_max_depth,
    { dispatchIndex, maxFanout: 8, seedKind, maxNodes: args.flow_max_nodes });
  bfsFlows = flowmod.suppressSubsets(bfsFlows);
  const dispatchFlows = flowmod.buildDispatchFlows(decls, dispatchIndex, edges, hubIds, { maxFanout: 8 });
  const flowList = [...dispatchFlows, ...bfsFlows];

  const ing = model.ingredients;
  const projectMeta = {
    name: ing.name,
    root: ing.root,
    languages: ing.languages,
    files_scanned: ing.files_scanned,
    files_by_language: ing.files_by_language,
    declarations_by_language: ing.declarations_by_language,
    parse_failures: ing.parse_failures,
    generated_at: new Date().toISOString().slice(0, 19),
  };
  projectMeta.template_detection = detection;
  projectMeta.resolution = ing.resolution;
  if (dispatchIndex.size) {
    projectMeta.dispatch = Object.fromEntries(
      [...dispatchIndex.entries()]
        .sort((a, b) => (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0))
        .map(([k, impls]) => [k, impls.map((d) => qualifiedName(d))]));
  }
  if (ing.advisories && ing.advisories.length) projectMeta.advisories = ing.advisories;
  if (ing.git) projectMeta.git = ing.git;
  if (ing.code_map_version) projectMeta.code_map_version = ing.code_map_version;
  if (ing.extract_version != null) projectMeta.extract_version = ing.extract_version;
  if (ing.refine_version != null) projectMeta.refine_version = ing.refine_version;

  return core.toJsonShape(
    decls, edges,
    layerLeaves.map((l) => ({ id: l.id, name: l.name, order: l.order,
      ...(l.summary_zh ? { summary_zh: l.summary_zh } : {}),
      ...(l.summary_en ? { summary_en: l.summary_en } : {}),
      ...(l.summary != null ? { summary: l.summary } : {}),
      ...(l.group ? { group: l.group } : {}) })),
    projectMeta, flowList, layerGroups);
}

export async function main(argv) {
  const args = parseArgs(argv);
  const ctx = buildContext(args);
  const root = ctx.root;
  let outPath = args.out;
  if (!isAbsolute(outPath)) outPath = join(root, outPath);
  mkdirSync(dirname(outPath), { recursive: true });
  const outDir = dirname(outPath);

  // --layer-only: consume a prior extract.json, run deterministic layering only.
  if (args.layer_only) {
    const extractPath = isAbsolute(args.extract) ? args.extract : join(root, args.extract);
    const model = deserializeExtract(JSON.parse(readFileSync(extractPath, 'utf8')));
    const data = runLayer(model, args, ctx);
    writeFileSync(outPath, JSON.stringify(data, null, 2));
    console.log(`[analyze] layer-only: ${data.layers.length} layers, ${data.flows.length} flows`);
    console.log(`[analyze] wrote ${outPath}`);
    return 0;
  }

  const model = await runExtract(args, ctx);

  // --extract-only: write the architecture-independent model + sidecars, stop.
  if (args.extract_only) {
    writeFileSync(outPath, JSON.stringify(serializeExtract(model), null, 2));
    writeUnresolved(model, outDir);
    writeResolution(model, outDir);
    console.log(`[analyze] root: ${root}`);
    console.log(`[analyze] languages: ${model.ingredients.languages.join(', ') || '(none)'}`);
    console.log(`[analyze] extract-only: ${model.declarations.length} declarations, ${model.edges.length} edges`);
    console.log(`[analyze] detection: ${model.detectionScores.chosen}`);
    console.log(`[analyze] wrote ${outPath} (extract.json — architecture-independent)`);
    return 0;
  }

  // Default (combined) — byte-identical to the legacy single-pass analyze.
  const data = runLayer(model, args, ctx);
  writeFileSync(outPath, JSON.stringify(data, null, 2));
  writeResolution(model, outDir);
  writeUnresolved(model, outDir);

  const det = data.project.template_detection;
  console.log(`[analyze] root: ${root}`);
  console.log(`[analyze] languages: ${model.ingredients.languages.join(', ') || '(none)'}`);
  if (det.reason) {
    console.log(`[analyze] template: ${det.chosen} (fallback: ${det.reason})`);
  } else {
    const ranked = Object.entries(det.scores).sort((x, y) => y[1] - x[1]).slice(0, 3)
      .map(([t, s]) => `${t}=${s}`).join(', ');
    console.log(`[analyze] template: ${det.chosen} (top: ${ranked})`);
  }
  console.log(`[analyze] files scanned: ${model.ingredients.files_scanned}  declarations: ${model.declarations.length}  edges: ${model.edges.length}`);
  if (det.fit && !det.fit.fits) {
    console.log(`[analyze] advisory: ${det.fit.warning} (Phase 2 should swap/derive layers)`);
  }
  console.log(`[analyze] flows: ${data.flows.length}`);
  console.log(`[analyze] wrote ${outPath}`);
  return 0;
}
